### PR TITLE
[SPARK-21898][ML] Feature parity for KolmogorovSmirnovTest in MLlib

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Feature parity for KolmogorovSmirnovTest in MLlib.
Implement `DataFrame` interface for `KolmogorovSmirnovTest` in `mllib.stat`.

## How was this patch tested?

Test suite added.

### CODE REVIEW

**Comment by WeichenXu123** [comment]
cc @yanboliang Thanks!

**Comment by kiszk** [comment]
Retest this please

**Comment by WeichenXu123** [comment]
@yanboliang

**Comment by WeichenXu123** [comment]
@sethah Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  /**
+   * Convenience function to conduct a one-sample, two-sided Kolmogorov-Smirnov test for probability
+   * distribution equality. Currently supports the normal distribution, taking as parameters
+   * the mean and standard deviation.
+   *
+   * @param dataset a `DataFrame` containing the sample of data to test
+   * @param sampleCol Name of sample column in dataset, of any numerical type
+   * @param distName a `String` name for a theoretical distribution, currently only support "norm".
+   * @param params `Double*` specifying the parameters to be used for the theoretical distribution
+   * @return DataFrame containing the test result for every feature against the label.
```
Comment (MrBago) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: I believe the input is a scalar column and the output is a double, is that right? We should update the language here to make that clear, it makes it sound like the inputs and outputs are Vectors.

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+   * Conduct the two-sided Kolmogorov-Smirnov (KS) test for data sampled from a
+   * continuous distribution. By comparing the largest difference between the empirical cumulative
+   * distribution of the sample data and the theoretical distribution we can provide a test for the
+   * the null hypothesis that the sample data comes from that theoretical distribution.
+   *
+   * @param dataset a `DataFrame` containing the sample of data to test
+   * @param sampleCol Name of sample column in dataset, of any numerical type
+   * @param cdf a `Double => Double` function to calculate the theoretical CDF at a given value
+   * @return DataFrame containing the test result for every feature against the label.
+   *         This DataFrame will contain a single Row with the following fields:
```
Comment (MrBago) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: Same as bellow.

**Review by MrBago — COMMENTED** [review]
@WeichenXu123 this lgtm, just a few minor comments.

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+   *
+   * @param dataset a `DataFrame` containing the sample of data to test
+   * @param sampleCol Name of sample column in dataset, of any numerical type
+   * @param cdf a `Double => Double` function to calculate the theoretical CDF at a given value
+   * @return DataFrame containing the test result for the input sampled data.
+   *         This DataFrame will contain a single Row with the following fields:
+   *          - `pValue: Double`
+   *          - `statistic: Double`
+   */
+  @Since("2.3.0")
```
Comment (MrBago) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: Sorry I didn't tag these in the first pass, can you update 2.4.

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+   * @param dataset a `DataFrame` containing the sample of data to test
+   * @param sampleCol Name of sample column in dataset, of any numerical type
+   * @param distName a `String` name for a theoretical distribution, currently only support "norm".
+   * @param params `Double*` specifying the parameters to be used for the theoretical distribution
+   * @return DataFrame containing the test result for the input sampled data.
+   *         This DataFrame will contain a single Row with the following fields:
+   *          - `pValue: Double`
+   *          - `statistic: Double`
+   */
+  @Since("2.3.0")
```
Comment (MrBago) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: Same.

**Comment by MrBago** [comment]
Thanks for the changes Weichen, this lgtm.

**Review by jkbradley — COMMENTED** [review]
Thanks for the PR!  I had a couple of small comments, plus a question about Java compatibility: Is the method taking a CDF Java-friendly?  I sent a Java test here: https://github.com/WeichenXu123/spark/pull/4

File: mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+ * @see <a href="https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test">
+ * Kolmogorov-Smirnov test (Wikipedia)</a>
+ */
+@Experimental
+@Since("2.4.0")
+object KolmogorovSmirnovTest {
+
+  /** Used to construct output schema of test */
+  private case class KolmogorovSmirnovTestResult(
+      pValues: Double,
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/stat/KolmogorovSmirnovTest.scala]: pValues -> pValue

**Comment by jkbradley** [comment]
LGTM
Thanks for the PR!
Merging with master

**Pull request closed without merging**